Properly document the local and global locality attributes. (#13831)
coq/coq: doc/sphinx/language/core/basic.rst

Locality attributes supported by :cmd:`Set` and :cmd:`Unset`
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The :cmd:`Set` and :cmd:`Unset` commands support the mutually
exclusive :attr:`local`, :attr:`export` and :attr:`global` locality

> attributes:

Reviewer: Do we need the full description of effects here? We could just reference to `_controlling-locality-of-commands` with a note that Set also has special alias syntax `Export`

Author: Indeed, now that we have a proper documentation of all locality attributes, we do not need to repeat it here (except for the weird default locality). Adding an explicit attribute for the default locality (as proposed in #13832) would even solve this remaining issue.

Author: I've applied your suggestion. Now this part of the text doesn't repeat the semantics of `local`, `export` and `global`.

* if no attribute is specified: the original value of the setting is
  restored at the end of the current module but it is *not* restored

>   at the end of the current section.

Author: No, see the example I added to #13832.

> * :attr:`local` (or the ``Local`` prefix): the setting

Reviewer 2: ```suggestion
* :attr:`local` (or the ``Local`` :n:`@legacy_attr`): the setting
```

and similar changes elsewhere

Author: I don't think it improves readability. Furthermore, there is the special case of `Export` which isn't a proper legacy attribute (as we've previously discussed) and which would break symmetry.

  is applied within the current module or section.  The original value
  of the setting is restored at the end of the current module or
  section.

> * :attr:`export` (or the ``Export`` prefix): same as no attribute, but
>   in addition, exported settings in :cmd:`Import`\ed modules override
>   the current value.

Reviewer 2: ```suggestion
* :attr:`export` (or the ``Export`` prefix): similar to no attribute, but
  settings exported from :cmd:`Import`\ed modules override
  the current value.
```

Or maybe "same as no attribute, except ...".  "In addition" is a weak way to convey the contradiction between the two clauses.  IMHO.

* :attr:`global` (or the ``Global`` prefix): the original value of the
  setting is *not* restored at the end of the current module or
  section.  In addition, global settings in a file override the

>   current value when the file is :cmd:`Require`\d (even transitively).

Reviewer 2: I notice you refer to files here and to modules above (line 538).  Is there a difference?  Maybe worth mentioning/repeating the distinction here (if there is one)?

Author: Yes, you can `Import` a module but only `Require` a file. Because a file is also a module, you can `Require Import` it. This whole chapter needs to be improved and consolidated anyway to make this fact clearer.

:attr:`global` and :attr:`export` locality attributes:

> * no attribute: the original setting is *not* restored at the end of

Reviewer 2: BTW, do we like the not restoring at the end of a section behavior or is it judged too disruptive to try to change it to something else?

  the current section but it is restored at the end of the current module.

> * :attr:`local` (or alternatively, the ``Local`` prefix): the setting

Author: I'd like to keep the current order for now, but I've changed the description of `export` to factorize more.

> * :attr:`local` (or alternatively, the ``Local`` prefix): the setting

Author: Now the descriptions of standard locality attributes `local`, `export` and `global` are not included there anymore.

  is applied within the current module or section.  The original value
  of the setting is restored at the end of the current module or
  section.
* :attr:`export` (or alternatively, the ``Export`` prefix): the
  original setting is *not* restored at the end of the current section
  but it is restored at the end of the current module. In addition,

>   the setting persists when :cmd:`Import`\ing the module in which it
>   is set.

Reviewer 2: ```suggestion
  exported settings in :cmd:`Import`\ed modules override the current value
```

* :attr:`global` (or alternatively, the ``Global`` prefix): the
  original setting is *not* restored at the end of the current module
  or section.  In addition, if the value is set in a file, then

>   :cmd:`Require`\-ing (even transitively) the file sets the option.

Reviewer 2: ```suggestion
  :cmd:`Require`\-ing the file (even transitively) sets the option.
```

The language in this section is confusing:
- we have "settings", which are flags and options, both of which can be "set" and "unset".
- "sets the option" doesn't seem to include flags
- if you unset a flag, can the "false" be exported?  If so, "if the value is set in a file" is misleading

I think we need a word that means "option or flag" other than "setting".  Or just use that phrase
repeatedly for clarity.  Maybe use "value" in place of some uses of "setting".

Author: >if you unset a flag, can the "false" be exported?

Yes.

Note that "setting" also includes table. Otherwise, option would be a generic term (flags are boolean options).

I will try to improve the wording.

Reviewer 2: Yes, "settings" includes tables, but the `Set` and `Unset` commands don't apply to tables, which use `Add` and `Remove` instead.  That's a little confusing.

Author: Indeed. And you just made me dig what was the behavior for `Add` and `Remove` and I was surprised to discover it is actually different (the default is `local` and `export` is not supported):

```coq
Section section.

  Add Search Blacklist "foo".

  Test Search Blacklist for "foo".
  (* Search blacklist does include foo *)

End section.

Test Search Blacklist for "foo".
(* Search blacklist does not include foo *)
```

Author: I've changed the text to use the wording "flag or option" instead of "setting".
